File: 1.Algorithm_Implementations/Sorting/Merge_Sort/JavaScript/mergeSort.js
function mergeSort(array) {
    sort(array, 0, array.length - 1);
    return array;
}

function sort(array, lowBound, highBound) {
    if (lowBound >= highBound) return;

    const midPoint = Math.floor(lowBound + (highBound - lowBound) / 2); // prevent's using float

    sort(array, lowBound, midPoint); // Left Partition
    sort(array, midPoint + 1, highBound); // Right Partition

    const auxArray = createAuxArray(array, lowBound, highBound, midPoint);
    copyAuxContent(array, auxArray, lowBound, highBound);
}

function createAuxArray(array, lowBound, highBound, midPoint) {
    const auxArray = new Array(highBound - lowBound + 1);
    let leftPointer = lowBound;
    let rightPointer = midPoint + 1;
    let auxIdx = 0;

    while (leftPointer <= midPoint && rightPointer <= highBound) {
        if (array[rightPointer] < array[leftPointer]) {
            auxArray[auxIdx++] = array[rightPointer++];
        } else {
            auxArray[auxIdx++] = array[leftPointer++];
        }
    }

    // Copy remaining left-side elements
    while (leftPointer <= midPoint) {
        auxArray[auxIdx++] = array[leftPointer++];
    }

    // Copy remaining right-side elements
    while (rightPointer <= highBound) {
        auxArray[auxIdx++] = array[rightPointer++];
    }

    return auxArray;
}

function copyAuxContent(array, auxArray, lowBound, highBound) {
    for (let i = lowBound; i <= highBound; i++) {
        array[i] = auxArray[i - lowBound];
    }
}

// Do not edit the line below. Needed for test cases
exports.mergeSort = mergeSort;

File: 1.Algorithm_Implementations/Sorting/Merge_Sort/Java/MergeSort.java

class MergeSort {
    public static int[] sort(int[] array) {
        sort(array, 0, array.length - 1);
        return array;
    }

    private static void sort(int[] array, int lowBound, int highBound) {
        if (lowBound >= highBound) {
            return;
        }

        int midPoint = lowBound + (highBound - lowBound) / 2;
        sort(array, lowBound, midPoint); // leftPartition
        sort(array, midPoint + 1, highBound); // rightPartition

        int[] auxArray = createAuxArray(array, lowBound, highBound, midPoint);
        copyAuxContent(array, auxArray, lowBound, highBound);
    }

    private static int[] createAuxArray(int[] array, int lowBound, int highBound, int midPoint) {
        int[] auxArray = new int[highBound - lowBound + 1];
        int leftPointer = lowBound;
        int rightPointer = midPoint + 1;
        int auxIdx = 0;

        while (leftPointer <= midPoint || rightPointer <= highBound) {
            if (leftPointer > midPoint || (rightPointer <= highBound && array[rightPointer] < array[leftPointer])) {
                auxArray[auxIdx++] = array[rightPointer++];
            } else {
                auxArray[auxIdx++] = array[leftPointer++];
            }
        }

        return auxArray;
    }

    private static void copyAuxContent(int[] array, int[] auxArray, int lowBound, int highBound) {
        for (int i = lowBound; i <= highBound; i++) {
            array[i] = auxArray[i - lowBound];
        }
    }

}
